mlflow/types/utils.py

```diff
@@ -79,7 +79,75 @@ def clean_tensor_type(dtype: np.dtype):
     return dtype
 
 
-def _infer_datatype(data) -> DataType:
+def _infer_colspec_type(data: Any) -> Union[DataType, Array, Object]:
+    """
+    Infer an MLflow Colspec type from the dataset.
+
+    :param data: data to infer from.
+    :return: Object
+    """
+    dtype = _infer_datatype(data)
+    _validate_inferred_type(dtype)
```

Comment:
I think we can remove this function and just raise exception here if dtype is None. Array, Object and Property all have dtype type check, so this `_validate_inferred_type` doesn't really validate anything except for `dypte=None`.

Reply:
You're right, I wasn't fully confident to add this `_validate_xxx` function only for `None` check. My only concern was to mistakenly allow some edge case in future edition, but on second thought, it can be caught by tests if there is. Will remove this func.